File: math/moments/C++/moments.cpp
#include <assert.h>
#include <cmath>
#include <functional>
#include <numeric>
#include <vector>

// Returns the arithmetic average of data
inline double ArithmeticAverage(const std::vector<double>& data)
{
	return std::accumulate(std::cbegin(data), std::cend(data), 0.0) / std::size(data);
}
// Returns the geometric average of data
inline double GeometricMean(const std::vector<double>& data)
{
	const double product{ std::accumulate(std::cbegin(data), std::cend(data), 1.0, std::multiplies<double>()) };
	return std::pow(product, 1.0 / std::size(data));
}
// Returns the harmonic average of data
inline double HarmonicMean(const std::vector<double>& data)
{
    const double sum{ 
        std::accumulate(
            std::cbegin(data), 
            std::cend(data), 
            0.0, 
            [](const double a, const double x){ return a + 1.0 / x; }
        ) 
    };
    return std::size(data) / sum;
}
// Helper to reduce against mean with an exponent
inline double Reduce(const std::vector<double>& data, const double mean, const int exponent)
{
    return std::accumulate(
        std::cbegin(data), 
        std::cend(data), 
        0.0, 
        [mean, exponent](const double a, const double x){ return a + std::pow(x - mean, exponent); }
    );
}
// Returns the unbiaised variance of data
inline double Variance(const std::vector<double>& data, const double mean)
{
	// divides by n-1 to return unbiaised estimation
	return  Reduce(data, mean, 2) / (std::size(data) - 1);
}
// Returns the skewness of data
inline double Skewness(const std::vector<double>& data, const double mean, const double variance)
{
	const size_t n{ std::size(data) };
	return n * Reduce(data, mean, 3) / std::pow(variance, 1.5) / (n - 1) / (n - 2);
}
// Returns the kurtosis of data
inline double Kurtosis(const std::vector<double>& data, const double mean, const double variance)
{
	const size_t n{ std::size(data) };
	const size_t nm1{ (n - 1) };
	const size_t nm2nm3{ (n - 2) * (n - 3) };
	const double first{ n * (n + 1) * Reduce(data, mean, 4) / nm1 / nm2nm3 / std::pow(variance, 2) };
	const double second{ 3.0 * nm1 * nm1 / nm2nm3 };
	return first - second;
}
// Returns the median of data
inline double Median(const std::vector<double>& data)
{
	size_t n{ std::size(data) / 2 };	
	std::vector<double> sortedData{ data };
	std::nth_element(std::begin(sortedData), std::begin(sortedData) + n, std::end(sortedData));
	return n % 2 == 0 ? sortedData[n] : 0.5 * (sortedData[n - 1] + sortedData[n]);
}

int main()
{
    const double tol{ 1.0e-05 };
    const std::vector<double> vec{ 1.2, 2.3, 3.4, 4.5, 5.6, 6.7, 7.8, 8.9, 9.0 };

	const double arithmeticMean{ ArithmeticAverage(vec) };
	assert(std::abs(arithmeticMean - 5.48889) < tol);
	assert(std::abs(GeometricMean(vec) - 4.636355) < tol);
	assert(std::abs(HarmonicMean(vec) - 3.652661) < tol);
	assert(std::abs(Median(vec) - 5.599999) < tol);
	
	const double variance{ Variance(vec, arithmeticMean) };
    assert(std::abs(variance - 8.086111) < tol);
    assert(std::abs(Skewness(vec, arithmeticMean, variance) - (-0.169877)) < tol);
    assert(std::abs(Kurtosis(vec, arithmeticMean, variance) - (-1.37685)) < tol);
}

File: math/moments/golang/moments.go
package main

//Using math and sort
import (
	"fmt"
	"math"
	"sort"
)

// Operation is the Binary Operation for func Accumulate()
type Operation func(float64, float64) float64

// Accumulate is the member function of []float64
func Accumulate(data []float64, initValue float64, f Operation) float64 {
	res := initValue
	for _, v := range data {
		res = f(res, v)
	}
	return res
}

// ArithmeticAverage returns the arithmetic average of data by Accumulate
func ArithmeticAverage(data []float64) float64 {
	return Accumulate(data, 0, func(a float64, b float64) float64 {
		return a + b
	}) / float64(len(data))
}

// GeometricMean returns the geometric average of data by Accumulate
func GeometricMean(data []float64) float64 {
	return math.Pow(Accumulate(data, 1, func(a float64, b float64) float64 {
		return a * b
	}), 1.0/float64(len(data)))
}

// HarmonicMean returns the harmonic average of data by Accumulate
func HarmonicMean(data []float64) float64 {
	return float64(len(data)) / Accumulate(data, 0, func(a float64, b float64) float64 {
		return a + 1.0/b
	})
}

// Reduce against mean with an exponent by Accumulate
func Reduce(data []float64, mean float64, exponent float64) float64 {
	return Accumulate(data, 0, func(a float64, b float64) float64 {
		return a + math.Pow(b-mean, exponent)
	})
}

// Variance returns unbiased variance of data by Reduce()
func Variance(data []float64, mean float64) float64 {
	return Reduce(data, mean, 2) / (float64(len(data)) - 1)
}

// Skewness returns skewness of data by Reduce()
func Skewness(data []float64, mean float64, variance float64) float64 {
	n := float64(len(data))
	return n * Reduce(data, mean, 3) / math.Pow(variance, 1.5) / (n - 1) / (n - 2)
}

// Kurtosis returns kurtosis of data by Reduce()
func Kurtosis(data []float64, mean float64, variance float64) float64 {
	n := float64(len(data))
	nm1 := n - 1
	nm2nm3 := (n - 2) * (n - 3)
	return n*(n+1)*Reduce(data, mean, 4)/nm1/nm2nm3/math.Pow(variance, 2) - 3.0*nm1*nm1/nm2nm3
}

// Median returns the median inside data
func Median(data []float64) float64 {
	size := len(data)
	sorted := data
	sort.Float64s(sorted)
	if size%2 == 0 {
		return (sorted[size/2-1] + sorted[size/2]) / 2.0
	}
	return sorted[size/2]
}

// Code to demonstrate
func main() {
	epilson := 1.0e-05
	vec := []float64{1.2, 2.3, 3.4, 4.5, 5.6, 6.7, 7.8, 8.9, 9.0}

	arithmeticMean := ArithmeticAverage(vec)
	fmt.Println(math.Abs(arithmeticMean-5.48889) < epilson)
	fmt.Println(math.Abs(GeometricMean(vec)-4.636355) < epilson)
	fmt.Println(math.Abs(HarmonicMean(vec)-3.652661) < epilson)
	fmt.Println(math.Abs(Median(vec)-5.599999) < epilson)

	variance := Variance(vec, arithmeticMean)
	fmt.Println(math.Abs(variance-8.086111) < epilson)
	fmt.Println(math.Abs(Skewness(vec, arithmeticMean, variance)-(-0.169877)) < epilson)
	fmt.Println(math.Abs(Kurtosis(vec, arithmeticMean, variance)-(-1.37685)) < epilson)
}
